components/uniset/src/builder.rs

```diff
@@ -0,0 +1,784 @@
+        let expected = vec![];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_remove_entire_range() {
+        let mut builder = generate_tester(vec![10, 20, 40, 50]);
+        builder.remove(10, 20);
+        let expected = vec![40, 50];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_remove_partial_range_left() {
+        let mut builder = generate_tester(vec![10, 20, 40, 50]);
+        builder.remove(10, 45);
+        let expected = vec![45, 50];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_remove_ne_range() {
+        let mut builder = generate_tester(vec![10, 20, 40, 50]);
+        builder.remove(20, 40);
+        let expected = vec![10, 20, 40, 50];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_remove_partial_range_right() {
+        let mut builder = generate_tester(vec![10, 20, 40, 50]);
+        builder.remove(15, 55);
+        let expected = vec![10, 15];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_remove_middle_range() {
+        let mut builder = generate_tester(vec![10, 20, 40, 50]);
+        builder.remove(12, 18);
+        let expected = vec![10, 12, 18, 20, 40, 50];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_remove_ne_middle_range() {
+        let mut builder = generate_tester(vec![10, 20, 40, 50]);
+        builder.remove(25, 27);
+        let expected = vec![10, 20, 40, 50];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_remove_encompassed_range() {
+        let mut builder = generate_tester(vec![10, 20, 40, 50, 70, 80]);
+        builder.remove(25, 55);
+        let expected = vec![10, 20, 70, 80];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_remove_char() {
+        let mut builder = generate_tester(vec![65, 70]);
+        builder.remove_char('A'); // 65
+        let expected = vec![66, 70];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_remove_range() {
+        let mut builder = generate_tester(vec![65, 90]);
+        builder.remove_range(&('A'..'L')); // 65 - 76
+        let expected = vec![76, 90];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_remove_set() {
+        let mut builder = generate_tester(vec![10, 20, 40, 50, 70, 80]);
+        let remove = UnicodeSet::from_inversion_list(vec![10, 20, 45, 75]).unwrap();
+        builder.remove_set(&remove);
+        let expected = vec![40, 45, 75, 80];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_retain_char() {
+        let mut builder = generate_tester(vec![65, 90]);
+        builder.retain_char('A'); // 65
+        let expected = vec![65, 66];
+        assert_eq!(builder.intervals, expected);
+    }
+
+    #[test]
+    fn test_retain_range() {
```

Comment:
Actually, for `retain_range` test, can you also include a scenario where the result is the empty set?  At least, there should be a test for a `retain_X` fn that returns an empty set, just to round out the test case scenarios.

The test for `retain_set` covers all of the interval overlap types and is testing a fn with similar but slightly less simple logic, so that's good.

Reply:
agreed and included (as `test_retain_range_empty()`)